### ISSUE TITLE
uri_string:parse returns bad error reason

### ISSUE THREAD

**Issue by schnittchen** [erlang/otp#5487]
**Describe the bug**

```
1> uri_string:parse("http://localhost:4000/session?foo[bar]=baz").
{error,invalid_uri,":"}
```

The error is the off-the-spec square brackets which should be quoted. The second third error triple value has no indication of this.

**Affected versions**
Erlang/OTP 24, possibly others

**Additional context**

See https://github.com/elixir-lang/elixir/issues/11450#issuecomment-986800174 for some context (that does not give much more info on the issue).

**Comment by zuiderkwast** [comment]
I'm trying to debug this. I don't have a solution but some findings.

This module has a lot code in the style of *try-catch-ignore-continue* which makes it hard to follow traces. I've added printouts at every `catch` clause to try to see what's happening:

```Erlang
45> c("src/uri_string").                                           
{ok,uri_string}
46> uri_string:parse("http://localhost:4000/session?foo[bar]=baz").
Caught {error,invalid_uri,"/"} in parse_hier
 at [{uri_string,parse_userinfo,2,[{file,"src/uri_string.erl"},{line,908}]},
     {uri_string,parse_hier,2,[{file,"src/uri_string.erl"},{line,839}]},
     {uri_string,parse_scheme,2,[{file,"src/uri_string.erl"},{line,811}]},
     {uri_string,parse_scheme_start,2,
                 [{file,"src/uri_string.erl"},{line,788}]},
     {uri_string,parse_uri_reference,2,
                 [{file,"src/uri_string.erl"},{line,626}]},
     {uri_string,parse,1,[{file,"src/uri_string.erl"},{line,366}]},
     {erl_eval,do_apply,6,[{file,"erl_eval.erl"},{line,684}]},
     {shell,exprs,7,[{file,"shell.erl"},{line,686}]}]
Caught {error,invalid_uri,"["} in parse_uri_reference
 at [{uri_string,parse_query,2,[{file,"src/uri_string.erl"},{line,1177}]},
     {uri_string,parse_segment,2,[{file,"src/uri_string.erl"},{line,724}]},
     {uri_string,parse_port,2,[{file,"src/uri_string.erl"},{line,1137}]},
     {uri_string,parse_reg_name,2,[{file,"src/uri_string.erl"},{line,998}]},
     {uri_string,parse_hier,2,[{file,"src/uri_string.erl"},{line,846}]},
     {uri_string,parse_scheme,2,[{file,"src/uri_string.erl"},{line,811}]},
     {uri_string,parse_scheme_start,2,
                 [{file,"src/uri_string.erl"},{line,788}]},
     {uri_string,parse_uri_reference,2,
                 [{file,"src/uri_string.erl"},{line,626}]}] 
Caught {error,invalid_uri,":"} in parse
 at [{uri_string,parse_segment_nz_nc,2,
                 [{file,"src/uri_string.erl"},{line,757}]},
     {uri_string,parse_relative_part,2,
                 [{file,"src/uri_string.erl"},{line,681}]},
     {uri_string,parse,1,[{file,"src/uri_string.erl"},{line,366}]},
     {erl_eval,do_apply,6,[{file,"erl_eval.erl"},{line,684}]},
     {shell,exprs,7,[{file,"shell.erl"},{line,686}]},
     {shell,eval_exprs,7,[{file,"shell.erl"},{line,642}]},
     {shell,eval_loop,3,[{file,"shell.erl"},{line,627}]}]
{error,invalid_uri,":"}
```

Explanation: After catching `{error,invalid_uri,"["}`, thrown in `parse_query` when `is_query(Char)` returns `false` and caught in `parse_uri_reference`, it does a fallback and tries to parse the whole URI as a path without scheme. Then it fails on the `:` in `http:`. This last error is what's returned in the end.

<details><summary>If you get error "Can't load module 'uri_string' that resides in sticky dir", do this</summary>

```Erlang
8> c("src/uri_string").                                           
=ERROR REPORT==== 27-Dec-2021::14:32:31.750013 ===
Can't load module 'uri_string' that resides in sticky dir

{error,sticky_directory}
9> code:which(uri_string).
"/home/viktor/repos/otp/lib/stdlib/uri_string.beam"
10> code:unstick_dir("/home/viktor/repos/otp/lib/stdlib").
ok
11> c("src/uri_string").                                  
{ok,uri_string}
```
</details>

**Comment by u3s** [comment]
thanks for sharing observation. I will try to have a look this week.

**Comment by u3s** [comment]
I agree with interpretation suggested by @zuiderkwast. In [GH-5487.txt](https://github.com/erlang/otp/files/7806389/GH-5487.txt), I've collected some printouts to have better view on what is happening. 

- line 53 is a place where unexpected '[' character is noticed by uri_string - this cause throwing an invalid_uri error for `[` character
- at this point library failed to parse an URI from received input
- uri_string does not give up yet, it catches the previous error and tries to parse input as `relative-ref`
- it fails again in line 59 and throws another invalid_uri error for `:` character, when trying to parse `path-noscheme`
- error comes from a `false` value returned by is_segment_nz_nc function (checking for non-zero segment without a colon character)

The error returned by uri_string:parse/1 represents the last error in the processing chain, it is a characteristic of current implementation.
I think behavior is also quite reasonable from RFC3986 perspective.

@schnittchen, @josevalim - please have a look and let me know if that explanation is sufficient.

**Comment by u3s** [comment]
Closing the issue. I think it works as designed in this case.
BTW, I recently found that that user guide is covering this example as well - https://www.erlang.org/doc/apps/stdlib/uri_string_usage.html#the-uri_string-module
